File: apple/HDHROpenKit/Sources/HDHROpenKit/ViewModels/GuideViewModel.swift
import Combine
import Foundation

@MainActor
public final class GuideViewModel: ObservableObject {
    @Published public private(set) var channels: [HDHomeRunChannel] = []
    @Published public private(set) var fullGuide: [HDHomeRunFullGuideChannel] = []
    @Published public private(set) var recordingRules: [HDHomeRunRecordingRule] = []
    @Published public private(set) var favoriteChannels: Set<String> = []
    @Published public var filterOnlyFavorites = false
    @Published public private(set) var isLoading = false
    @Published public private(set) var guideAvailable = false
    @Published public private(set) var error: String?

    public let apiClient: APIClient
    private let watchSessionManager: WatchSessionManager

    public init(apiClient: APIClient, watchSessionManager: WatchSessionManager) {
        self.apiClient = apiClient
        self.watchSessionManager = watchSessionManager
    }

    public var displayedChannels: [HDHomeRunChannel] {
        if filterOnlyFavorites, !favoriteChannels.isEmpty {
            return channels.filter { favoriteChannels.contains($0.channelNumber) }
        }
        return channels
    }

    public func loadData() async {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        defer { isLoading = false }

        async let channelsTask = loadChannels()
        async let guideTask = loadGuide()
        async let favoritesTask = loadFavorites()
        async let rulesTask = loadRules()

        _ = await (channelsTask, guideTask, favoritesTask, rulesTask)
    }

    public func loadChannels() async {
        do {
            let resp = try await apiClient.getChannels()
            channels = resp.channels
            guideAvailable = resp.guideAvailable
        } catch {
            self.error = error.localizedDescription
            Log.general.error("Failed to load channels: \(error.localizedDescription)")
        }
    }

    public func loadGuide() async {
        do {
            let now = Date().timeIntervalSince1970
            let guide = try await apiClient.getGuide(start: now - 6 * 3600, end: now + 48 * 3600)
            fullGuide = guide
        } catch {
            Log.general.warning("Failed to load full guide: \(error.localizedDescription)")
        }
    }

    public func loadFavorites() async {
        do {
            let integration = try await apiClient.getNetworkIntegration(type: "hdhomerun")
            if let favs = integration.settings["favorite_channels"]?.value as? [String] {
                favoriteChannels = Set(favs)
            }
        } catch {
            // Favoriting is optional
        }
    }

    public func loadRules() async {
        do {
            recordingRules = try await apiClient.listRecordingRules()
        } catch {
            Log.dvr.warning("Failed to load recording rules: \(error.localizedDescription)")
        }
    }

    public func toggleFavorite(channelNumber: String) async {
        var updated = favoriteChannels
        if updated.contains(channelNumber) {
            updated.remove(channelNumber)
        } else {
            updated.insert(channelNumber)
        }
        favoriteChannels = updated

        // Persist to backend network integration settings
        do {
            let favsArray = Array(updated)
            _ = try await apiClient.updateNetworkIntegration(
                type: "hdhomerun",
                settings: ["favorite_channels": AnyCodable(favsArray)]
            )
        } catch {
            Log.general.error("Failed to save favorite channels: \(error.localizedDescription)")
        }
    }

    public func findRule(for channelNumber: String?, airing: HDHomeRunGuideEntry?) -> HDHomeRunRecordingRule? {
        RecordingRuleMatcher.findMatchingRule(rules: recordingRules, channelNumber: channelNumber, airing: airing)
    }

    public func recordEpisode(
        seriesId: String? = nil,
        channelNumber: String? = nil,
        start: TimeInterval? = nil,
        options: RecordingRuleOptions? = nil
    ) async throws {
        let payload = AddRecordingRulePayload(
            seriesId: (seriesId?.isEmpty ?? true) ? "auto" : seriesId!,
            dateTime: start,
            channel: options?.channel ?? channelNumber,
            title: options?.title,
            titleMatchMode: options?.titleMatchMode,
            keywordQuery: options?.keywordQuery,
            recentOnly: options?.recentOnly,
            startPadding: options?.startPadding,
            endPadding: options?.endPadding,
            maxEpisodesToKeep: options?.maxEpisodesToKeep,
            server: options?.server
        )
        recordingRules = try await apiClient.addRecordingRule(payload: payload)
        await loadRules()
    }

    public func recordSeries(
        seriesId: String,
        channelNumber: String? = nil,
        options: RecordingRuleOptions? = nil
    ) async throws {
        let payload = AddRecordingRulePayload(
            seriesId: seriesId.isEmpty ? "auto" : seriesId,
            channel: options?.channel ?? channelNumber,
            title: options?.title,
            titleMatchMode: options?.titleMatchMode,
            keywordQuery: options?.keywordQuery,
            recentOnly: options?.recentOnly ?? false,
            startPadding: options?.startPadding,
            endPadding: options?.endPadding,
            maxEpisodesToKeep: options?.maxEpisodesToKeep,
            server: options?.server
        )
        recordingRules = try await apiClient.addRecordingRule(payload: payload)
        await loadRules()
    }

    public func updateRule(
        ruleId: String,
        isSeries: Bool,
        options: RecordingRuleOptions? = nil,
        seriesId: String? = nil,
        start: TimeInterval? = nil,
        channelNumber: String? = nil
    ) async throws {
        let payload = AddRecordingRulePayload(
            seriesId: (seriesId?.isEmpty ?? true) ? "auto" : seriesId!,
            dateTime: isSeries ? nil : start,
            channel: options?.channel ?? channelNumber,
            title: options?.title,
            titleMatchMode: options?.titleMatchMode,
            keywordQuery: options?.keywordQuery,
            recentOnly: isSeries ? (options?.recentOnly ?? false) : nil,
            startPadding: options?.startPadding,
            endPadding: options?.endPadding,
            maxEpisodesToKeep: options?.maxEpisodesToKeep,
            server: options?.server
        )
        recordingRules = try await apiClient.updateRecordingRule(id: ruleId, payload: payload)
        await loadRules()
    }

    public func cancelRule(ruleId: String) async throws {
        recordingRules = try await apiClient.deleteRecordingRule(id: ruleId)
        await loadRules()
    }

    public func getAirings(for channelNumber: String) -> [HDHomeRunGuideEntry] {
        if let full = fullGuide.first(where: { $0.channelNumber == channelNumber }) {
            return full.airings
        }
        if let ch = channels.first(where: { $0.channelNumber == channelNumber }) {
            var items: [HDHomeRunGuideEntry] = []
            if let now = ch.now {
                items.append(now)
            }
            if let next = ch.next {
                items.append(next)
            }
            return items
        }
        return []
    }
}

File: android/core/src/main/kotlin/org/hdhropen/kit/viewmodels/GuideViewModel.kt
package org.hdhropen.kit.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonPrimitive
import org.hdhropen.kit.models.*
import org.hdhropen.kit.networking.APIClient
import org.hdhropen.kit.utilities.Log
import org.hdhropen.kit.utilities.RecordingRuleMatcher

class GuideViewModel(
    val apiClient: APIClient
) : ViewModel() {
    private val _channels = MutableStateFlow<List<HDHomeRunChannel>>(emptyList())
    val channels: StateFlow<List<HDHomeRunChannel>> = _channels.asStateFlow()

    private val _fullGuide = MutableStateFlow<List<HDHomeRunFullGuideChannel>>(emptyList())
    val fullGuide: StateFlow<List<HDHomeRunFullGuideChannel>> = _fullGuide.asStateFlow()

    private val _recordingRules = MutableStateFlow<List<HDHomeRunRecordingRule>>(emptyList())
    val recordingRules: StateFlow<List<HDHomeRunRecordingRule>> = _recordingRules.asStateFlow()

    private val _favoriteChannels = MutableStateFlow<Set<String>>(emptySet())
    val favoriteChannels: StateFlow<Set<String>> = _favoriteChannels.asStateFlow()

    val filterOnlyFavorites = MutableStateFlow(false)

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _guideAvailable = MutableStateFlow(false)
    val guideAvailable: StateFlow<Boolean> = _guideAvailable.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    val displayedChannels: List<HDHomeRunChannel>
        get() {
            val all = _channels.value
            val favs = _favoriteChannels.value
            if (filterOnlyFavorites.value && favs.isNotEmpty()) {
                return all.filter { favs.contains(it.channelNumber) }
            }
            return all
        }

    fun loadData() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            try {
                val channelsDef = async { loadChannels() }
                val guideDef = async { loadGuide() }
                val favsDef = async { loadFavorites() }
                val rulesDef = async { loadRules() }

                awaitAll(channelsDef, guideDef, favsDef, rulesDef)
            } finally {
                _isLoading.value = false
            }
        }
    }

    suspend fun loadChannels() {
        try {
            val resp = apiClient.getChannels()
            _channels.value = resp.channels
            _guideAvailable.value = resp.guideAvailable
        } catch (e: Exception) {
            _error.value = e.localizedMessage
            Log.general.error("Failed to load channels: ${e.localizedMessage}")
        }
    }

    suspend fun loadGuide() {
        try {
            val guide = apiClient.getGuide()
            _fullGuide.value = guide
        } catch (e: Exception) {
            Log.general.warning("Failed to load full guide: ${e.localizedMessage}")
        }
    }

    suspend fun loadFavorites() {
        try {
            val integration = apiClient.getNetworkIntegration("hdhomerun")
            val favsElement = integration.settings["favorite_channels"]
            if (favsElement is JsonArray) {
                val favs = favsElement.mapNotNull {
                    if (it is JsonPrimitive && it.isString) it.content else null
                }.toSet()
                _favoriteChannels.value = favs
            }
        } catch (e: Exception) {
            // Favoriting is optional
        }
    }

    suspend fun loadRules() {
        try {
            _recordingRules.value = apiClient.listRecordingRules()
        } catch (e: Exception) {
            Log.dvr.warning("Failed to load recording rules: ${e.localizedMessage}")
        }
    }

    fun toggleFavorite(channelNumber: String) {
        viewModelScope.launch {
            val current = _favoriteChannels.value.toMutableSet()
            if (current.contains(channelNumber)) {
                current.remove(channelNumber)
            } else {
                current.add(channelNumber)
            }
            _favoriteChannels.value = current

            try {
                val jsonArray = JsonArray(current.map { JsonPrimitive(it) })
                apiClient.updateNetworkIntegration(
                    type = "hdhomerun",
                    settings = mapOf("favorite_channels" to jsonArray)
                )
            } catch (e: Exception) {
                Log.general.error("Failed to save favorite channels: ${e.localizedMessage}")
            }
        }
    }

    fun findRule(channelNumber: String?, airing: HDHomeRunGuideEntry?): HDHomeRunRecordingRule? {
        return RecordingRuleMatcher.findMatchingRule(_recordingRules.value, channelNumber, airing)
    }

    suspend fun recordEpisode(
        seriesId: String? = null,
        channelNumber: String? = null,
        start: Double? = null,
        options: RecordingRuleOptions? = null
    ) {
        val payload = AddRecordingRulePayload(
            seriesId = if (seriesId.isNullOrEmpty()) "auto" else seriesId,
            dateTime = start,
            channel = options?.channel ?: channelNumber,
            title = options?.title,
            titleMatchMode = options?.titleMatchMode,
            keywordQuery = options?.keywordQuery,
            recentOnly = options?.recentOnly,
            startPadding = options?.startPadding,
            endPadding = options?.endPadding,
            maxEpisodesToKeep = options?.maxEpisodesToKeep,
            server = options?.server
        )
        _recordingRules.value = apiClient.addRecordingRule(payload)
        loadRules()
    }

    suspend fun recordSeries(
        seriesId: String,
        channelNumber: String? = null,
        options: RecordingRuleOptions? = null
    ) {
        val payload = AddRecordingRulePayload(
            seriesId = seriesId.ifEmpty { "auto" },
            channel = options?.channel ?: channelNumber,
            title = options?.title,
            titleMatchMode = options?.titleMatchMode,
            keywordQuery = options?.keywordQuery,
            recentOnly = options?.recentOnly ?: false,
            startPadding = options?.startPadding,
            endPadding = options?.endPadding,
            maxEpisodesToKeep = options?.maxEpisodesToKeep,
            server = options?.server
        )
        _recordingRules.value = apiClient.addRecordingRule(payload)
        loadRules()
    }

    suspend fun updateRule(
        ruleId: String,
        isSeries: Boolean,
        options: RecordingRuleOptions? = null,
        seriesId: String? = null,
        start: Double? = null,
        channelNumber: String? = null
    ) {
        val payload = AddRecordingRulePayload(
            seriesId = if (seriesId.isNullOrEmpty()) "auto" else seriesId,
            dateTime = if (isSeries) null else start,
            channel = options?.channel ?: channelNumber,
            title = options?.title,
            titleMatchMode = options?.titleMatchMode,
            keywordQuery = options?.keywordQuery,
            recentOnly = if (isSeries) (options?.recentOnly ?: false) else null,
            startPadding = options?.startPadding,
            endPadding = options?.endPadding,
            maxEpisodesToKeep = options?.maxEpisodesToKeep,
            server = options?.server
        )
        _recordingRules.value = apiClient.updateRecordingRule(ruleId, payload)
        loadRules()
    }

    suspend fun cancelRule(ruleId: String) {
        _recordingRules.value = apiClient.deleteRecordingRule(ruleId)
        loadRules()
    }

    fun getAirings(channelNumber: String): List<HDHomeRunGuideEntry> {
        val full = _fullGuide.value.firstOrNull { it.channelNumber == channelNumber }
        if (full != null && full.airings.isNotEmpty()) {
            return full.airings
        }

        val ch = _channels.value.firstOrNull { it.channelNumber == channelNumber }
        if (ch != null) {
            val list = mutableListOf<HDHomeRunGuideEntry>()
            ch.now?.let { list.add(it) }
            ch.next?.let { list.add(it) }
            return list
        }
        return emptyList()
    }
}
